KEP-3027: SLSA Level 3 Compliance in the Kubernetes Release Process (#3051)
kubernetes/enhancements: keps/sig-release/3027-slsa-compliance/README.md

[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This document proposes a plan to harden the Kubernetes releases by
making the necessary adjustments to comply with the SLSA Framework.
[SLSA (Supply-chain Levels for Software Artifacts)](https://slsa.dev/)
is a framework to harden software supply currently being defined by the 
[OpenSSF](https://openssf.org/)'s 
[Supply Chain Integrity WG](https://github.com/ossf/wg-supply-chain-integrity).

The framework provides requirements and recommendations to software
build systems to harden their environments and the processes that drive
them. It also defines the metadata that needs to be produced to trace the 
origins of every item in a software release.

The main goal of this enhancement is to provide downstream consumers of our
artifacts the highest assurance about the integrity of each Kubernetes release.

SLSA defines several levels of hardening, each touching more aspects of the
release process that go beyond its technical implementation. This document is
meant to serve as a guide to reach the highest possible levels after 

> consensus has been reached about their viability.

Reviewer: This is a KEP, the consensus will be reached before merging it. So I would rephrase to be more concrete:

> This KEP proposes to use SLSA as a standard framework for hardening the Kubernetes release software supply chain.

Author: Sounds good. I've reworded it with your suggestion and added the next point:

> <!--

Reviewer: I would additionally state that we're not only proposing SLSA for k/k artifacts, but also as a standard recommendation (no must) for every other project in the kubernetes and kubernetes-sigs org.

Author: .. added a line around this. Recommending and also sharing tools and practices as we learn them :+1:

This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

## Motivation

> Release Kubernetes in a zero-trust environment.

Reviewer: I would make this either a headline or rephrase for a higher verbosity, like:

> The main motivation for reaching full SLSA compliance is to be able to release Kubernetes in a zero-trust environment.

Author: OK, changed it :)

Kubernetes releases represent key links in many software supply chains, not
just for the project itself but also for consumers that derive, repackage and
distribute our artifacts downstream. The project releases end-user artifacts
like binaries and container images, but also source code that is actively reused
further down the distribution stream.

All current work done by SIG Release on the Kubernetes supply chain centers
around three focus areas: Artifact Consumption, Introspection, and Security.
These areas are explained in more detail in our [_Roadmap for 2021 and Beyond_ 

> document](https://github.com/kubernetes/sig-release/blob/master/roadmap.md).

Reviewer: Link the doc to the latest `master` commit please.

Author: Done :+1:

As the world hardens its software distribution methods, the Kubernetes project needs
to do its part to achieve a secure supply chain from end to end: from the
top base images to the final artifacts downloaded by end users. This proposal
provides a path to achieve increasing levels of security, integrity, and availability
in our releases by engineering new features to our processes. The objective is
to achieve the highest possible compliance with the

> [SLSA framework](https://slsa.dev/) (Supply-chain Levels for Software Artifacts).

Reviewer: We already define the abbreviation in the Summary section.
```suggestion
[SLSA framework](https://slsa.dev/).
```

in our roadmap: Artifacts can be consumed easier and with more trust.
Improvements to code and process will secure the supply chain and each release
will produce software bills of materials, provenance attestations and signatures
which will yield much better introspection to the journey from code to binary.

The SLSA framework is a project under the [OpenSSF](https://openssf.org/)'s [Digital Identity 
Attestation Working Group](https://github.com/ossf/wg-digital-identity-attestation).
The framework defines numbered levels of compliance that harden software supply
chains by recommending concrete steps to address, each of increasing technical
complexity, number of stakeholders, and process adjuments necessary to reach them. 

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

> The main goal of this KEP is to provide a plan to achieve the highest possible
> SLSA compliance in the Kubernetes release process. While providing a roadmap

Reviewer: As already mentioned in the non-goals below, we do not provide a plan. An incomplete plan maybe, but I'd like to rephrase it more to state that the main goal of this kep is the highest SLSA compliance level for k/k, for example:
```suggestion
The main goal of this KEP is to define the highest possible
SLSA compliance as final objective in enhancing the Kubernetes release process. While providing a roadmap
```

this KEP can be considered complete when a SLSA level is deemed as not
implementable by the community at large (see Graduation Criteria).

> 

Reviewer: I'd like to add as a goal that we track the progress of the SLSA compliance effort as part of our documentation.

Author: Good point. Addedd a paragraph adding full documentation as a goal, including specs and handbook updates.

Kubernetes release.

#### Provenance Metadata

Once the KEP is implemented, end-users will be able to verify the provenance of
the artifacts we release. This means that the origin of each artifact will be
traceable to its precise origin in the build process, with all parameters,
inputs, code points, and other metadata available in a verifiable and 
non-repudiable format.    

#### Release Completeness

Signed supply chain metadata files, like the software bill of materials and the
provenance attestations, will allow downstream consumers to ensure the integrity and
completeness of a Kubernetes release. This means users can be sure that all expected
artifacts are there, untampered, and also check that no extra items filtered into
the release buckets, registries, etc.

<!--
### Notes/Constraints/Caveats (Optional)

SLSA requirements involve changes to software that releases Kubernetes, to 
-->

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

> Failure to pursue this enhancement poses three problems for the project:

Reviewer: I only see two :upside_down_face: 
```suggestion
Failure to pursue this enhancement poses two problems for the project.
```

- Will there be e2e and integration tests, in addition to unit tests?
- How will it be tested in isolation vs with other components?

No need to outline all of the test cases, just the general strategy. Anything
that would count as tricky in the implementation, and anything particularly
challenging to test, should be called out.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

### Graduation Criteria

This KEP can be considered complete when one of two scenarios is reached:

1. All SLSA levels have been successfully complied with 
1. The community determines one of the levels as not implementable. This
may be concluded if the nature of necessary changes proves to be too
disruptive or implies altering aspects of technical environments
and/or systems, contributor experience, policy, and other domains beyond what
deems to be acceptable.

Tracking issues will be opened to track and discuss the viability of the

> required enhancements to reach each SLSA level.

Reviewer: Let's add that we also update the roadmap to reflect the current state.

Author: OK, added.

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This document proposes a plan to harden the Kubernetes releases by
making the necessary adjustments to comply with the SLSA Framework.
[SLSA (Supply-chain Levels for Software Artifacts)](https://slsa.dev/)

> is a framework to harden software supply chains, currently in actove

Reviewer 2: ```suggestion
is a framework to harden software supply chains, currently in active
```

Author: Thanks @JTarasovic !
